core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -582,18 +584,33 @@ import scala.util.control.NonFatal
                                 exception: Exception): Unit = {
           // this is invoked on the thread calling consumer.poll which will always be the actor, so it is safe
           val duration = System.nanoTime() - startTime
-          if (duration > settings.commitTimeWarning.toNanos) {
-            log.warning("Kafka commit took longer than `commit-time-warning`: {} ms, commitsInProgress={}",
-                        duration / 1000000L,
-                        commitsInProgress)
-          }
           commitsInProgress -= 1
-          if (exception != null) {
-            val failure = Status.Failure(exception)
-            replyTo.foreach(_ ! failure)
-          } else {
-            commitRefreshing.committed(offsets)
-            replyTo.foreach(_ ! Done)
+          exception match {
+            case null =>
+              if (duration > settings.commitTimeWarning.toNanos) {
+                log.warning("Kafka commit took longer than `commit-time-warning`: {} ms, commitsInProgress={}",
+                            duration / 1000000L,
+                            commitsInProgress)
+              }
+              commitRefreshing.committed(offsets)
+              replyTo.foreach(_ ! Done)
+
+            case e: RetriableCommitFailedException =>
+              log.warning("Kafka commit is to be retried, after={} ms, commitsInProgress={}, cause={}",
+                          e.getCause,
+                          duration / 1000000L,
+                          commitsInProgress)
+              commitMaps = commitMap.toList ++ commitMaps
+              commitSenders = commitSenders ++ replyTo
+              requestDelayedPoll()
+
+            case commitException =>
+              log.warning("Kafka commit failed after={} ms, commitsInProgress={}, exception={}",
```

Comment:
Will returning this exception to the asker fail the stream?  Maybe it should be logged as an error if it's not already done later.

Reply:
Yes, that's the exception that fails the stream. But if the user doesn't properly listen to stream completion it will get lost.

Reply:
I changed it to log as error.